redhat-appstudio/remote-secret
# [mshaposh]  upload/validation webhook

### What does this PR do?

- Adds ability to upload secretData along with RemoteSecret creating/updating using Mutating Webhook.
- Validates the upload secret for targets uniqueness


### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-569

It significantly improves ability of dynamic SecretData uploading. Comparing to using Upload Secret (which is NOT excluded anyway) it:

increases security of secret data (it never appeared in etcd)
improves UX and supportability, using sync REST call in a webhook allows reacting immediately (user creating the Secret will get clear error message instead of analyzing error CR status


### How to test this PR?

Build/push/deploy as usual.
1) Upload (mutating) webhook
Put SecretData to the extra-field RemoteSecret->data as an array of key/value pairs, like:

```
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: myrs1
  namespace: default
data:
  username: Z2VuYQ==
  password: Z2VuYQ==
```

Create RemoteSecret and check:

RS called myrs1 is created on default namespace and does not contain 'data' field


2) Validating webhook

Try to create a RS with duplicating targets, liike:

```
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  name: myrs2
  namespace: default
spec:
  targets:
    - namespace: ns
    - namespace: ns
```    
Validating error shows up, RS is not created
